### PR TITLE
Implement parameterisable modules

### PR DESCRIPTION
**Pull request by thiolliere**
related to https://github.com/paritytech/substrate/issues/1688

This PR contains currently the modification of Balances (and so srml-support macros) to be able to be used with an instance parameter representing the actual instance. Thus dependant module could make use of two balances at the same time or runtime could use many balances etc...

#### For module using Balances:

This balance module has been modified but default to a specific instance thus it almost doesn't break API for existing module depending on this one. Also the prefix used for storage of default instance is same as old one. The only difference is GenesisConfig that has a new PhantomData field.

#### For Balances module:

Not that much change is needed just declarations have new parameters and some inner struct have also an Instance generic.
And also it declares its trait with a generic parameter: `Trait<Instance: Instantiable=DefaultInstance>` (note: the bound to Instantiable is not required)
use can just write `Trait<Instance=DefaultInstance>`

#### Macros:

decl_storage: The whole difficulty is being able to implement `fn key() -> &'static [u8]`.
>(how did I get there)
>First I thought about implementing storages for some concrete types U0 to U16 for example but that would have required a lot of where clause in decl_storage `impl` module but also in user `impl`.
>Secondly I thought about make Instance implementing a trait: `Trait Instantiable { const PREFIX: &'static str; }`  using lazy_static to create a final key that is the concatenation of old prefix (cratename + storagename) and a number:
>```rust
>fn key() -> &'static [u8] {
>lazy_static!{ static PREFIX: String = { 
>    let a = String::from(#prefix);
>    a.push_str(Instance::PREFIX);
>    a
>};
>}
>```
> but that fails to compile because of static of Instance being used in outer function.

The final solution is derived from it: I actually create the trait Instantiable inside decl_storage with the methods I need and implement it for some structure I make. Expanded it results in
```rust
pub trait Instantiable {
  const PREFIX: &'static str:
  build_prefix_once_for_AStorageName(suffix: &'static [u8]) -> &'static [u8];
  build_prefix_once_for_AnotherStorageName(suffix: &'static [u8]) -> &'static [u8];
}
struct DefaultInstance;
impl Instantiable for DefaultInstance {
  const PREFIX: &'static str=""// to be compatible with old prefix
  build_prefix_once_for_AStorageName(suffix: &'static [u8]) -> &'static [u8] {
	static LAZY: #scrate::lazy::Lazy<Vec<u8>> = #scrate::lazy::Lazy::INIT;
	LAZY.get(|| {
		let mut final_prefix = Vec::new();
		final_prefix.extend_from_slice(suffix);
		final_prefix.extend_from_slice(Self::PREFIX.as_bytes());
		final_prefix
	})

  }
  build_prefix_once_for_AnotherStorageName(suffix: &'static [u8]) -> &'static [u8] {
	static LAZY: #scrate::lazy::Lazy<Vec<u8>> = #scrate::lazy::Lazy::INIT;
	LAZY.get(|| {
		let mut final_prefix = Vec::new();
		final_prefix.extend_from_slice(suffix);
		final_prefix.extend_from_slice(Self::PREFIX.as_bytes());
		final_prefix
	})
  }
}
```
As a note methods couldn't be implemented in a generic way because of how static is build by rust. To understand why see https://github.com/hukumka/generic_static

and then key function is implemented this way:
```rust
fn key() -> &'static [u8] {
Instance::build_prefix_once_for_....(#old_prefix)
}
```
and the key will be computed only once.

decl_storage: modified to be able to have an instance and a default instance on its structure (module, call_type etc...)

decl_event: no that much things changed just being able to make Event generic over two parameters if wanted.

#### What to do:

* support linked map.
* macro call are not very good and inherited from previous attempt. we could simplify things. omit the default parameter as it is not configuration or make it configuratable, make instance just a keyword with the number of instance wanted. I'm on it.
* code should be rewritten in some parts.
* documentation.
* tests

also as a note: a multiple instance module requires now a a decl_storage to actually create the trait.

also we have to keep in mind this might complexify decl_storage in way we don't want. But all in all it seems we always can just make the trait requiring methods we need.

### CODE REVIEW

File: srml/support/procedural/src/storage/impls.rs [hunk 1]
```diff
@@ -215,166 +249,189 @@ impl<'a> Impls<'a> {
...
-							res.1.previous = self.previous;
-							storage.put(&*next_key, &res);
-						}
-					}
+		// generator for linked map
+		//
+		// NOTE: we split implementation because of recursion limit
+		let mut impls = quote! {
+
+			#[derive(Default, #scrate::parity_codec_derive::Encode, #scrate::parity_codec_derive::Decode)]
```
Comment (tomusdrw) [inline srml/support/procedural/src/storage/impls.rs]: We should implement `Derive` manually to avoid unnecessary bound on `Key`

File: srml/support/procedural/src/storage/impls.rs [hunk 2]
```diff
@@ -215,166 +249,189 @@ impl<'a> Impls<'a> {
...
+				///
+				/// Takes care of updating previous and next elements points
+				/// as well as updates head if the element is first or last.
+				fn remove<S: #scrate::GenericStorage>(
+					linkage: #linkage<#kty>,
+					storage: &S,
+				) {
+					use #scrate::storage::generator::StorageMap;
+
+					let next_key = linkage.next.as_ref().map(<Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for);
```
Comment (tomusdrw) [inline srml/support/procedural/src/storage/impls.rs]: Can we avoid repetition of this `<Self as #scrate...StorageMa..>`, it's really terrible. maybe do `let as_map = quote! { .. };` ?

File: srml/balances/src/lib.rs [hunk 3]
```diff
@@ -156,6 +156,7 @@ decl_storage! {
 		pub ReservedBalance get(reserved_balance): map T::AccountId => T::Balance;
 
 
+		Dummy2: linked_map T::Balance => T::Balance;
```
Comment (gavofyork) [inline srml/balances/src/lib.rs]: ??

**Comment by tomusdrw** [comment]
There is also quite a lot of todos in the code.

File: srml/democracy/src/lib.rs [hunk 4]
```diff
@@ -492,6 +492,7 @@ mod tests {
 			transfer_fee: 0,
 			creation_fee: 0,
 			vesting: vec![],
+			_genesis_phantom_data: rstd::marker::PhantomData,
```
Comment (gavofyork) [inline srml/democracy/src/lib.rs]: This is pretty horrible...

**Comment by gavofyork** [comment]
> There is also quite a lot of todos in the code.

To be fair, it's marked `inprogress` :)

File: srml/democracy/src/lib.rs (hunk 4 above)
Comment (thiolliere) [reply srml/democracy/src/lib.rs]: indeed,
When GenesisConfig needs to access to storage it needs to be generic other the trait. This trait is sometimes contained in GenesisConfig fields and sometimes not, in the latter cases we already add this field _genesis_phantom_data.

Now I changed the condition to requires phantom data also when the trait is generic over instances.
But this condition does actually have false positive. For instance the balances module could also not have a phantom, the Instance generic type is used in fields like T::Balance.
But as we don't access Trait definition inside the macro there is no way to know if T::Balance is actually an associated type from Trait (and so use the Instance generic type parameter) or an associated type from a dependency of Trait like T::AccountId would be (and so doesn't use Instance generic type parameter).

To summarize: This field is not always needed, it is not for balances, but there is no way to know inside the macro. The only solution I see is: we could provide a way for the user to say no_phantom_data (this seems good and would make balances module truly backward compatible)

Also note that user could write instead:
```rust
		t.extend(GenesisConfig::<Test>{
			dummy: 42,
			foo: 24,
			.. Default::default()
		}.build_storage().unwrap().0);
```

**Comment by thiolliere** [comment]
small update, also I'll wait for https://github.com/paritytech/substrate/pull/1809 and then impl for linked_map.

* user of Balance don't have anything to change now as a new keyword has been added to the decl_storage macro to be able to specify to skip phantom data field when we know it is not necessary.

**Comment by thiolliere** [comment]
small update: implemented for linked_map, still have to do documentation and test but otherwise eveything seems fine to me.

**Comment by thiolliere** [comment]
working the construct_runtime it will probably end up to something like 
`Example0: example::{Module<example::Instance0>, Event<T, example::Instance0>}`

also as a note: I wonder in how many situation this would be useful. If we want to implement it for balance at the end then it worth implementing this way. But if we only expect some user to implement it for a few cases then I don't know if it worth maintaining this way. Because one could simply do a macro like this:
```rust
multiple_module! {
  pub trait Trait: system::Trait {
  }
  decl_module! {
  ...
}
```
with multiple_module macro expanding to
```rust
mod module1 {
  pub trait Trait: system::Trait {
  }
  decl_module! {
  ...
}

mod module2 {
  pub trait Trait: system::Trait {
  }
  decl_module! {
  ...
}
```
And this would be equivalent in terms of feature for other modules and runtime.

**Comment by thiolliere** [comment]
@gavofyork actually I wonder about your opinions on this feature, mine are mixed.
The good part is that it is entirely backward compatible it only extend definiton of modules. the bad part is that it add a bit of complexity on our side for a need I don't know how to estimate.
But if the need is here. It could be finished by the end of the week.

**Comment by xlc** [comment]
Some feedback from a parachain developer.
We are working on a chain with multiple tokens and we have a module implements all the tokens. It provides multiple implementation of the `Currency` trait so we can pass different `Currency` trait to different modules.
We don't need multiple instantiations of a module, we need multiple implementations of a trait, which is already supported.
The general trend of runtime modules should be depending on traits, instead of depending on a particular module.

**Comment by thiolliere** [comment]
thanks @xlc 
as note I could imagine a use case like a module requiring two currency:
```rust
trait Trait {
  type Currency1: Currency;
  type Currency2: Currency;
}
```
a user of this module could use balance for each currency but with two different instance
```rust
impl Trait for Runtime {
  type Currency1 = balance::Module<Self, balance::Instance1>;
  type Currency2 = balance::Module<Self, balance::Instance2>;
}
```

**Comment by thiolliere** [comment]
small update: I am more confident in the code now, still have to check Log part and inherent part.
Also the new syntax is implemented you can take a look at node/runtime as I temporarly pushed the test:
```rust
construct_runtime!(
	pub enum Runtime with Log(InternalLog: DigestItem<Hash, SessionKey>) where
		Block = Block,
		NodeBlock = node_primitives::Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: system::{default, Log(ChangesTrieRoot)},
....
		Balances1: balances::<Instance1>::{Event<T, I>},
		Example: example::<Instance1>::{Module, Call, Storage, Event<T, I>, Config<T, I>, Origin<T, I>},
	}
);
```

**Comment by thiolliere** [comment]
full example builds, inherent, origin and log are implemented.

it requires parity-codec to implement Codec for core::marker::Phantom

**Comment by thiolliere** [comment]
### Current status:
* This PR contains a full implementation and a modification of node/runtime to test compilation and show usage (not meant to be merge in)
* To test this for non regression I think it is better to build a test runtime which I don't know to do at the moment.

### Details on implementation
* A lot of macros used to parse entry has been rewritten in a more compact way. The reason for it was actually to be able to implement it faster: instead of explicitly rewritte all parsed and rest types at each step of this parsing we take advantage of token hygien type `tt`. Those token are but inside {} resulting in something like:
```rust
	(
// prelude
		$runtime:ident;
// parsed
		{ $( $parsed:tt )* };
// actual parsing
		$name:ident: $module:ident:: $( < $module_instance:ident >:: )? {
			$ingore:ident $( <$ignor:ident $(, $ignore_instance:path)?> )* $(, $modules:ident $( <$modules_generic:ident $(, $modules_instance:path)?> )* )*
		},
// rest
		$( $rest:tt )*
	) => {
		$crate::__decl_outer_config!(
			$runtime;
			{ $( $parsed )* };
			$name: $module:: $( < $module_instance >:: )? { $( $modules $( <$modules_generic $(, $modules_instance)?> )* ),* },
			$( $rest )*
		);
	};
```

### How is decl_storage implemented
* it creates an Instantiable trait that provides the methods to get prefix or else:
```rust
pub trait Instantiable {
    const INSTANCE_PREFIX: &'static str;
    fn build_prefix_once_for_Dummy(prefix: &'static [u8]) -> &'static [u8];
    fn build_prefix_once_for_Bar(prefix: &'static [u8]) -> &'static [u8];
    fn build_head_key_once_for_Bar(prefix: &'static [u8]) -> &'static [u8];
    fn build_prefix_once_for_Foo(prefix: &'static [u8]) -> &'static [u8];
}

impl Instantiable for Instance10 {
	const INSTANCE_PREFIX: &'static str = "Instance10";
	fn build_prefix_once_for_Dummy(prefix: &'static [u8]) -> &'static [u8] {
		static LAZY: lazy::Lazy<Vec<u8>> = lazy::Lazy::INIT;
		LAZY.get(|| {
			let mut final_prefix = Vec::new();
			final_prefix.extend_from_slice(prefix);
			final_prefix.extend_from_slice(Self::INSTANCE_PREFIX.as_bytes());
			final_prefix
		})
	}
...
}

// then in implementation
// we call the or build it as such:

let final_prefix = if let Some(instance) = instance {
	let method_name = syn::Ident::new(&format!("build_prefix_once_for_{}", name.to_string()), proc_macro2::Span::call_site());
	quote!{ #instance::#method_name(#prefix.as_bytes()) }
} else {
	quote!{ #prefix.as_bytes() }
};
                                                                                                                                 
// make sure to use different prefix for head and elements.
let final_head_key = if let Some(instance) = instance {
	let method_name = syn::Ident::new(&format!("build_head_key_once_for_{}", name.to_string()), proc_macro2::Span::call_site());
	quote!{ #instance::#method_name(#prefix.as_bytes()) }
} else {
	let final_head_key = format!("head of {}", prefix);
	quote!{ #final_head_key.as_bytes() }
};
```


### Notes:

* All in all this PR just had the possibility of an additional generic for module, storages, event, log, origin, call and modifies the macros as such.
* This can add a bit of complexity on the decl_storage implementation and some other macros
* If someone want to reuse a module (that doesn't implements Instance) he can also create a new module sharing 99% of the code and changing names.
* If someone does want to implement multiple instance of a module without this PR he can also make the module as a template and write the correct derivation in build.rs (though debugging may not be handy I don't know)

File: srml/support/test/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+[package]
```
Comment (bkchr) [inline srml/support/test/Cargo.toml]: Why do we need an extra crate for this?

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: I written it in the commit message:

this allow to test for construct_runtime as well.

The reason to have put that in another crate is:
* if we put test in `tests/` dir of srml/support then decl_storage fails to get
  srml-support access because it believes it is inside srml-support
  crate and so derive access to `quote!{ crate }` but this is wrong
  (and I don't see any way to prevent that, and it only bother us so I
  don't think that matters that much)
* if we put test inside lib.rs then contruct_runtime cannot be used
  because it call some macros that are defined with macros
  (decl_outer_event and decl_outer_origin) and thus rustc complains.

**Comment by thiolliere** [comment]
a small update on it. It doesn't support having a module with instances and event without instance generic parameter. This may be solved by making an event per instance or just enforce event to be generic over instances.

this is because: impl_outer_event, we need to be able to convert event from module instance to outer event
```rust
		$(
			impl From<$module_name::Event $( <$generic_param $(, $generic_instance)? > )*> for $name {
				fn from(x: $module_name::Event $( <$generic_param $(, $generic_instance)? > )*) -> Self {
					$name::$module_name(x)
				}
			}
		)*
```

File: srml/support/test/Cargo.toml (hunk 5 above)
Comment (bkchr) [reply srml/support/test/Cargo.toml]: Yeah, just fix the `crate` access. Instead of generating `crate`, use the hidden include. While generating the hidden include, change the generation to check if `CARGO_PKG_NAME` is `srml-support` and import `srml-support` by `extern crate srml-support` instead of using my `proc_macro_crate`. Just look into `parity-codec`.

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: Yes we do so already:
```rust
	let scrate = generate_crate_access(&hidden_crate_name, "srml-support");
```
```rust 
/// Generates the access to the `srml-support` crate.
pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
	if ::std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
		quote::quote!( crate )
	} else {
		let mod_name = generate_hidden_includes_mod_name(unique_id);
		quote::quote!( self::#mod_name::hidden_include )
	}.into()
}
```

`CARGO_PKG_NAME` is `srml-support` for test in tests dir.
I don't know how we can determine that we are actually in this kind of environment.
in std::env current_dir return "/home/thiolliere/Developpement/substrate/srml/support" and args are "/home/thiolliere/Developpement/substrate/target/debug/deps/instances-edf4e94ffdc37966"

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: oh actually it seems we can:

when CARGO_PKG_NAME == srml-support && arg dir doesn't start with "srml_support-"

Comment (bkchr) [reply srml/support/test/Cargo.toml]:
```
pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
	let mod_name = generate_hidden_includes_mod_name(unique_id);
	quote::quote!( self::#mod_name::hidden_include )
}
```
+ 
```
pub fn generate_hidden_includes(unique_id: &str, def_crate: &str) -> TokenStream {
	if ::std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
		quote::quote!(
			#[doc(hidden)]
			mod #mod_name {
				pub extern crate srml-support as hidden_include;
			}
		)
	} else {
		let mod_name = generate_hidden_includes_mod_name(unique_id);

		match crate_name(def_crate) {
			Ok(name) => {
				let name = Ident::new(&name, Span::call_site());
				quote::quote!(
					#[doc(hidden)]
					mod #mod_name {
						pub extern crate #name as hidden_include;
					}
				)
			},
			Err(e) => {
				let err = Error::new(Span::call_site(), &e).to_compile_error();
				quote!( #err )
			}
		}

	}.into()
}
```

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: But then inlined test inside srml/support/src can't work anymore because they can't write `extern crate srml_support as hidden_include;`

Comment (bkchr) [reply srml/support/test/Cargo.toml]: Hmm yeah :( With `1.34` we can start using `extern crate self`. So continue using an extra crate :)

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: oh cool!

Comment (thiolliere) [reply srml/support/test/Cargo.toml]: yes we still won't be able to write `extern crate srml_support as hidden_include;`
in inline test of srml-support.
and extern crate self as hidden_include in tests dir doesn't link to srml-support but to the test itself.

**Comment by thiolliere** [comment]
OK, so now tests are somewhat complete, I'm still not sure about this line https://github.com/paritytech/substrate/pull/1800/files#diff-a35cd0dc28321418bffa3a01045354deR252

But I think it is ready for review

**Comment by thiolliere** [comment]
OK I'm still not sure about this https://github.com/paritytech/substrate/pull/1800/files#diff-a35cd0dc28321418bffa3a01045354deR282:

```rust
		pub type Event<$event_generic_param $(, $instance $( = $event_default_instance)? )?> = RawEvent<$( $generic_type ),* $(, $instance)? >;
		pub enum RawEvent<$( $generic_param ),* $(, $instance)? > {
			#[doc(hidden)]
			$(PhantomData($crate::rstd::marker::PhantomData<$instance>),)?
			$(
				$events
			)*
		}
```

When creating event with a generic modules it needs to be generic over the trait and instance in order (to be able to convert into outer event).
Sometimes the event use I (instance generic parameter) as it uses associated types of Trait. But sometimes it only use associated type of dependency of Trait like balances::Trait and else and then it must add a phantom variant that uses the generic parameter.

It can be let to the user to create this variant when needed. I don't know what is better.

**Comment by thiolliere** [comment]
needs a new minor release of parity-codec before merge and then ok

**Comment by gavofyork** [comment]
@thiolliere is the minor release something you can take care of?

**Comment by gavofyork** [comment]
i'l like to push forward and get this in now.

**Comment by bkchr** [comment]
Wouldn't it be easier and much less inversive to add the instance as associated type?

**Comment by thiolliere** [comment]
@bkchr we can't implement multiple times the same trait with a different associated type, either we make the trait generic or we provide multiple traits.

note: indeed providing multiple traits could be a solution, I would implement it with a global procedural macro like
```rust
// expand to multiple modules or a single module but different trait like TraitInstance2, TraitInstance3
// rename storage name for each implementation as $(ModuleName)Instance2, ModuleNameInstance3
decl_multiple_instance! {
// whole module except imports
}
```

**Comment by thiolliere** [comment]
ok codec published, dependency updated, and master merged
